В данном файле буду рассматривать исключительно трешхолд 20 дней

In [1]:
import pandas as pd
data = pd.read_excel('Arvand.xlsx')
data

KOD  Birthday  Birthday1    Year  Married   Gender Nationality  \
0       70081389      1975 1925-02-05  1975.0  Оиладор  Мужской       Узбек   
1       70068103      1933 1933-03-27     NaN  Оиладор  Мужской       Точик   
2       70084699      1933 1933-03-27     NaN  Оиладор  Мужской       Точик   
3       70067500      1933 1933-11-27     NaN   Беоила  Женский       Точик   
4        1047372      1938 1938-04-23  1938.0  Оиладор  Мужской       Узбек   
...          ...       ...        ...     ...      ...      ...         ...   
85417  210106137      1984        NaT  1984.0  Оиладор  Мужской       Узбек   
85418  210107051      1990        NaT  1990.0   Беоила  Мужской       Узбек   
85419  210108260      1974        NaT  1974.0  Оиладор  Женский       Узбек   
85420  210108426      1985        NaT  1985.0  Оиладор  Мужской       Узбек   
85421  210108465      1980        NaT  1980.0  Оиладор  Женский       Узбек   

                Educ FamilySize BusExper  ...  \
0                Оли          3       15  ...   
1              Миёна          3      NaN  ...   
2              Миёна          3      NaN  ...   
3      Миёнаи махсус   Нет инфо        3  ...   
4              Миёна          1        5  ...   
...              ...        ...      ...  ...   
85417          Миёна   Нет инфо       10  ...   
85418          Миёна          1      NaN  ...   
85419            Оли          3      NaN  ...   
85420          Миёна          2        5  ...   
85421          Миёна          2        4  ...   

      Залог (группа-поручительство, категория 1 - недвижимость, категория 2-движимое имущество, категория 3-поручительство, категория 4 - без залога)  \
0                                            Категория 2                                                                                                
1                                            Категория 4                                                                                                
2                                            Категория 3                                                                                                
3                                            Категория 2                                                                                                
4                                            Категория 4                                                                                                
...                                                  ...                                                                                                
85417                                        Категория 2                                                                                                
85418                                        Категория 2                                                                                                
85419                                        Категория 2                                                                                                
85420                                        Категория 3                                                                                                
85421                                             Группа                                                                                                

       Капитал    Актив Обязательства Дни просрочки  \
0       2820.0   3000.0           0.0            46   
1          0.0      0.0           0.0            11   
2          0.0      0.0           0.0             0   
3       5050.0   5200.0           0.0             0   
4       6777.0   7600.0         543.0            15   
...        ...      ...           ...           ...   
85417      0.0      0.0           0.0            53   
85418   5000.0   5000.0           0.0             0   
85419  51600.0  96400.0       44800.0             0   
85420  12600.0  12600.0           0.0             2   
85421  13880.0  15080.0        12

# 1)Предобработка данных

### Удаление ненужных ячеек

In [2]:
data1=data.copy()

In [3]:
# Создаем новый столбец 'Combined_Birthday'
data1['Combined_Birthday'] = pd.NaT  # Инициализируем пустыми значениями

# Цикл для создания новой даты и заполнения столбца 'Combined_Birthday'
for index, row in data1.iterrows():
    #Берем год рождения из столбца Birthday, если ячейка пустая, то в качестве года рождения устанавливаем 1960г, т.е. возраст 54-55 лет
    YY = row['Birthday'] if not pd.isnull(row['Birthday']) else 1960
    #Берем день рождения из столбца Birthday1, если ячейка пустая, то в качестве дня рождения устанавливаем 1
    dd = row['Birthday1'].day if not pd.isnull(row['Birthday1']) else 1
    #Берем месяц рождения из столбца Birthday1, если ячейка пустая, то в качестве месяца рождения устанавливаем январь
    mm = row['Birthday1'].month if not pd.isnull(row['Birthday1']) else 1
    #Заполняем новый столбец датой рождения в формата %d-%m-%Y
    combined_date = pd.to_datetime(f"{dd}-{mm}-{YY}", errors='coerce')
    data1.at[index, 'Combined_Birthday'] = combined_date


C:\Users\extre\AppData\Local\Temp\ipykernel_14536\3181191886.py:13: UserWarning: Parsing dates in %d-%m-%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  combined_date = pd.to_datetime(f"{dd}-{mm}-{YY}", errors='coerce')


In [4]:
#Удаляем явно ненужные столбцы
data1 = data1.drop(['KOD', 'Birthday1', 'Year', 'Birthday','Максимальные дни просрочки'], axis=1)

In [5]:
data1

Married   Gender Nationality           Educ FamilySize BusExper  \
0      Оиладор  Мужской       Узбек            Оли          3       15   
1      Оиладор  Мужской       Точик          Миёна          3      NaN   
2      Оиладор  Мужской       Точик          Миёна          3      NaN   
3       Беоила  Женский       Точик  Миёнаи махсус   Нет инфо        3   
4      Оиладор  Мужской       Узбек          Миёна          1        5   
...        ...      ...         ...            ...        ...      ...   
85417  Оиладор  Мужской       Узбек          Миёна   Нет инфо       10   
85418   Беоила  Мужской       Узбек          Миёна          1      NaN   
85419  Оиладор  Женский       Узбек            Оли          3      NaN   
85420  Оиладор  Мужской       Узбек          Миёна          2        5   
85421  Оиладор  Женский       Узбек          Миёна          2        4   

           Filial         Region  \
0      Истаравшан      Шахристон   
1      Истаравшан      Гули сурх   
2      Истаравшан      Гули сурх   
3      Истаравшан      Шахристон   
4          Хучанд  Худжанд-Центр   
...           ...            ...   
85417  Ч. Расулов          Куруш   
85418  Ч. Расулов       Спитамен   
85419  Ч. Расулов     Гулякандоз   
85420  Ч. Расулов     Дж.Расулов   
85421  Ч. Расулов     Дж.Расулов   

      Уровень клиента (Бовари до 1 год, Хамкори от 1 до 3 лет, Шарик от 3 до 5 лет, VIPболее 5 лет)  \
0                                                 Бовари                                              
1                                                 Бовари                                              
2                                                Хамкори                                              
3                                                 Бовари                                              
4                                                Хамкори                                              
...                                                  ...                                              
85417                                             Бовари                                              
85418                                            Хамкори                                              
85419                                                VIP                                              
85420                                              Шарик                                              
85421                                              Шарик                                              

        Навпреление деятельности  ...  \
0      Чорводори ва коркарди шир  ...   
1           Приобретение техники  ...   
2                    Ремонт дома  ...   
3      Чорводори ва коркарди шир  ...   
4      Чорводори ва коркарди шир  ...   
...                          ...  ...   
85417  Чорводори ва коркарди шир  ...   
85418                  Кишту кор  ...   
85419                Ремонт дома  ...   
85420  Чорводори ва коркарди шир  ...   
85421                      Савдо  ...   

      Залог (группа-поручительство, категория 1 - недвижимость, категория 2-движимое имущество, категория 3-поручительство, категория 4 - без залога)  \
0                                            Категория 2                                                                                                
1                                            Категория 4                                                                                                
2                                            Категория 3                                                                                                
3                                            Категория 2                                                                                                
4                                            Категория 4                                                                                                
...                                  

In [6]:
print(data1['Дата выдачи'])

0       2015-01-19
1       2014-03-05
2       2015-04-01
3       2014-02-21
4       2014-01-13
           ...    
85417   2015-07-16
85418   2015-09-02
85419   2015-11-17
85420   2015-11-12
85421   2015-11-11
Name: Дата выдачи, Length: 85422, dtype: datetime64[ns]


In [7]:
import numpy as np
# Создание нового столбца "Age". От даты выдачи кредита отнимаем год рождения.
#Так как результат выводит только количество дней, поэтому делим на 365.25, чтобы учивать также высокосные годы
data1['Age'] = (data1['Дата выдачи'] - data1['Combined_Birthday']).dt.days / 365.25


In [8]:
#Все значения столбца Age делаем целочисенным
data1.Age = data1['Age'].astype(int)

In [9]:
data1.head()

Married   Gender Nationality           Educ FamilySize BusExper  \
0  Оиладор  Мужской       Узбек            Оли          3       15   
1  Оиладор  Мужской       Точик          Миёна          3      NaN   
2  Оиладор  Мужской       Точик          Миёна          3      NaN   
3   Беоила  Женский       Точик  Миёнаи махсус   Нет инфо        3   
4  Оиладор  Мужской       Узбек          Миёна          1        5   

       Filial         Region  \
0  Истаравшан      Шахристон   
1  Истаравшан      Гули сурх   
2  Истаравшан      Гули сурх   
3  Истаравшан      Шахристон   
4      Хучанд  Худжанд-Центр   

  Уровень клиента (Бовари до 1 год, Хамкори от 1 до 3 лет, Шарик от 3 до 5 лет, VIPболее 5 лет)  \
0                                             Бовари                                              
1                                             Бовари                                              
2                                            Хамкори                                              
3                                             Бовари                                              
4                                            Хамкори                                              

    Навпреление деятельности  ... Капитал   Актив Обязательства Дни просрочки  \
0  Чорводори ва коркарди шир  ...  2820.0  3000.0           0.0            46   
1       Приобретение техники  ...     0.0     0.0           0.0            11   
2                Ремонт дома  ...     0.0     0.0           0.0             0   
3  Чорводори ва коркарди шир  ...  5050.0  5200.0           0.0             0   
4  Чорводори ва коркарди шир  ...  6777.0  7600.0         543.0            15   

   Количество просрочек  Этап кредитования (который раз получает кредит  \
0                     4                                               1   
1                     5                                               1   
2                     0                                               2   
3                     0                                               1   
4                     2                                               2   

  Валовая прибыль  Чистая прибыль  Combined_Birthday Age  
0             0.0           180.0         1975-05-02  39  
1             0.0             0.0         1933-03-27  80  
2             0.0          1648.0         1933-03-27  82  
3             0.0           150.0         1933-11-27  80  
4             0.0           280.0         1938-04-23  75  

[5 rows x 30 columns]

In [10]:
#Удаляем столбец Combined_Birthday, так как он больше нам не нужен
data1=data1.drop('Combined_Birthday', axis=1)

In [11]:
data1.head()

Married   Gender Nationality           Educ FamilySize BusExper  \
0  Оиладор  Мужской       Узбек            Оли          3       15   
1  Оиладор  Мужской       Точик          Миёна          3      NaN   
2  Оиладор  Мужской       Точик          Миёна          3      NaN   
3   Беоила  Женский       Точик  Миёнаи махсус   Нет инфо        3   
4  Оиладор  Мужской       Узбек          Миёна          1        5   

       Filial         Region  \
0  Истаравшан      Шахристон   
1  Истаравшан      Гули сурх   
2  Истаравшан      Гули сурх   
3  Истаравшан      Шахристон   
4      Хучанд  Худжанд-Центр   

  Уровень клиента (Бовари до 1 год, Хамкори от 1 до 3 лет, Шарик от 3 до 5 лет, VIPболее 5 лет)  \
0                                             Бовари                                              
1                                             Бовари                                              
2                                            Хамкори                                              
3                                             Бовари                                              
4                                            Хамкори                                              

    Навпреление деятельности  ...  \
0  Чорводори ва коркарди шир  ...   
1       Приобретение техники  ...   
2                Ремонт дома  ...   
3  Чорводори ва коркарди шир  ...   
4  Чорводори ва коркарди шир  ...   

  Залог (группа-поручительство, категория 1 - недвижимость, категория 2-движимое имущество, категория 3-поручительство, категория 4 - без залога)  \
0                                        Категория 2                                                                                                
1                                        Категория 4                                                                                                
2                                        Категория 3                                                                                                
3                                        Категория 2                                                                                                
4                                        Категория 4                                                                                                

  Капитал   Актив Обязательства  Дни просрочки  Количество просрочек  \
0  2820.0  3000.0           0.0             46                     4   
1     0.0     0.0           0.0             11                     5   
2     0.0     0.0           0.0              0                     0   
3  5050.0  5200.0           0.0              0                     0   
4  6777.0  7600.0         543.0             15                     2   

  Этап кредитования (который раз получает кредит  Валовая прибыль  \
0                                              1              0.0   
1                                              1              0.0   
2                                              2              0.0   
3                                              1              0.0   
4                                              2              0.0   

   Чистая прибыль Age  
0           180.0  39  
1             0.0  80  
2          1648.0  82  
3           150.0  80  
4           280.0  75  

[5 rows x 29 columns]

#### Преобразование категориальных переменных в фиктивные переменные (переменные со значениеми 0 и 1 и т.д.) 

In [12]:
data1.columns

Index(['Married', 'Gender', 'Nationality', 'Educ', 'FamilySize', 'BusExper',
       'Filial', 'Region',
       'Уровень клиента (Бовари до 1 год, Хамкори от 1 до 3 лет, Шарик от 3 до 5 лет, VIPболее 5 лет)',
       'Навпреление деятельности', 'Продукт', 'Сектор', 'Дата выдачи',
       'Валюта кредита', 'Сумма кредита', 'Срок кредита',
       'Дата окончания по графику', 'Ежемесячная сумма погашения по графику',
       'Льготный период (месяц)',
       'Залог (группа-поручительство, категория 1 - недвижимость, категория 2-движимое имущество, категория 3-поручительство, категория 4 - без залога)',
       'Капитал', 'Актив', 'Обязательства', 'Дни просрочки',
       'Количество просрочек',
       'Этап кредитования (который раз получает кредит', 'Валовая прибыль',
       'Чистая прибыль', 'Age'],
      dtype='object')

In [13]:
data2 = data1.drop(['Уровень клиента (Бовари до 1 год, Хамкори от 1 до 3 лет, Шарик от 3 до 5 лет, VIPболее 5 лет)',
                    'Дата выдачи','Дата окончания по графику', 'Капитал', 'Актив','Валовая прибыль'], axis=1)

In [14]:
data2.columns

Index(['Married', 'Gender', 'Nationality', 'Educ', 'FamilySize', 'BusExper',
       'Filial', 'Region', 'Навпреление деятельности', 'Продукт', 'Сектор',
       'Валюта кредита', 'Сумма кредита', 'Срок кредита',
       'Ежемесячная сумма погашения по графику', 'Льготный период (месяц)',
       'Залог (группа-поручительство, категория 1 - недвижимость, категория 2-движимое имущество, категория 3-поручительство, категория 4 - без залога)',
       'Обязательства', 'Дни просрочки', 'Количество просрочек',
       'Этап кредитования (который раз получает кредит', 'Чистая прибыль',
       'Age'],
      dtype='object')

In [15]:
data2.Married.unique()

array(['Оиладор', 'Беоила', 'Бевамард (бевазан)', 'Чудошуда'],
      dtype=object)

In [16]:
#Используем метод Ordinal encoding так как есть последовательность
married_mapping = {'Оиладор': 0, 'Беоила': 1, 'Бевамард (бевазан)': 2, 'Чудошуда': 3}
data2['FamilyStatus'] = data2['Married'].map(married_mapping)

In [17]:
data2.Gender.unique()

array(['Мужской', 'Женский'], dtype=object)

In [18]:
#Используем метод Ordinal encoding так как есть последовательность
gender_mapping = {'Мужской': 0, 'Женский': 1}
data2['Gender'] = data2['Gender'].map(gender_mapping)

In [19]:
for column in data1.columns:
    print(column, data1[column].unique())

Married ['Оиладор' 'Беоила' 'Бевамард (бевазан)' 'Чудошуда']
Gender ['Мужской' 'Женский']
Nationality ['Узбек' 'Точик' 'Тотор' 'Рус' 'Киргиз' 'Украин' 'Другие' 'Карис'
 'Карачои']
Educ ['Оли' 'Миёна' 'Миёнаи махсус' 'Олии нопурра' 'Миёнаи нопурра'
 'Аспирантура']
FamilySize [3 'Нет инфо' 1 2 6 8 4 5 7 nan]
BusExper [15 nan 3 5 25 10 39 30 20 40 6 16 18 2 19 23 14 45 17 35 1 4 49 46 50 37
 7 31 8 12 33 34 9 26 38 21 22 42 24 36 28 11 44 41 32 47 0 48 13 43 '20+'
 27 29 75 60 ',' 521 'ъолия' 100 2.5 1.5 3.5 4.5 1.4 1.1 2.4 '6мох' 0.6
 'мир' '6-мох' '7-мох' '4ё' 123 2.3 '0.5' 0.8 141 1.3 '8-мох' 0.7 0.5 215
 113 4.6 2.7 73 '4-мох' '6мес' 1.6 6.5 '5 мох' 1.7 '1мох' 0.4]
Filial ['Истаравшан' 'Хучанд' 'Ч. Расулов' 'Душанбе' 'Исфара' 'Панчакент']
Region ['Шахристон' 'Гули сурх' 'Худжанд-Центр' 'Спитамен' 'Шарк' 'Мархамат'
 'Душанбе' 'Навкент' 'Кистакуз' 'Худжанд-Панчшанбе' 'Бустон'
 'Истаравшан-филиал' 'Рудаки' 'Ашт' 'Калининобод' 'Сино' 'Исфара' 'Хисор'
 'Зафаробод' 'Ничони' 'Вахдат' 'Мехнат

In [20]:
#Используем метод Ordinal encoding так как есть последовательность
educ_mapping = { 'Аспирантура': 0, 'Оли': 1, 'Миёнаи махсус': 2, 'Олии нопурра': 3, 'Миёна': 4, 'Миёнаи нопурра': 5}
data2['Education'] = data2['Educ'].map(educ_mapping)

In [21]:
#Кодируем 'FamilySize' для 'Нет инфо' и заполняем пустые значения нулями
data2.loc[data2['FamilySize'] == 'Нет инфо', 'FamilySize'] = 0
data2['FamilySize'].fillna(0, inplace=True)

In [22]:
# Кодировка столбца 'BusExper' с помощью `map`
bus_map = {
    ',': 0, 0: 0, 0.4: 0, '0.5': 0, 0.5: 0, 0.6: 0, 0.7: 0, 0.8: 0, 1: 0, 1.1: 0, 1.3: 0, 1.4: 0, 1.5: 0, 1.6: 0, 1.7: 0, '1мох': 0, 2: 0,
    2.3: 0, 2.4: 0, 2.5: 0, 2.7: 0, 3: 0, 3.5: 0, 4: 0, '4-мох': 0, 4.5: 0, 4.6: 0, 5: 0, '5 мох': 0, 'мир': 0,
    6: 1, '6-мох': 0, 6.5: 1, '6мес': 0, '6мох': 0, 7: 1, '7-мох': 0, 8: 1, '8-мох': 0, 9: 1, 10: 1, 100: 1,
    11: 2, 113: 2, 123: 2, 12: 2, 13: 2, 14: 2, 15: 2,
    16: 3, 17: 3, 18: 3, 19: 3, 20: 3,
    '20+': 4, 21: 4, 215: 4, 22: 4, 23: 4, 24: 4, 25: 4,
    26: 5, 27: 5, 28: 5, 29: 5, 30: 5, 31: 5, 32: 5, 33: 5, 34: 5, 35: 5, 36: 5, 37: 5, 38: 5, 39: 5, 40: 5,
    141: 6, 41: 6, 42: 6, 43: 6, 44: 6, 45: 6, 46: 6, 47: 6, 48: 6, 49: 6, '4ё': 6, 50: 6,
    521: 7, 73: 7, 75: 7, 'ъолия': 7
}

# Создаем новый столбец и присваиваем ему кодированные значения
data2['business_experience'] = data2['BusExper'].map(bus_map)


Я разделил BusExper на интервалы 0-5 лет стажа, которое принимает метку 0, 5-10 лет - метка 1, 10-15 лет - метка 2, 15-20 лет стажа - метка 3, 20-25 лет - метка 4, 25-40 лет - метка 5, 40-50 лет стажа - метка 6, 50+ лет стажа - метка 7. При этом я предполагал, что значения больше 100 были введени случайно, и присваивал каждый к интервалу, которые больше всего подходит

In [23]:
#Используем Label Encoder, чтобы уменьшить количество столбцов
data2.loc[(data2['Продукт']=='Потребительский кредит') |
          (data2['Продукт']=='Жилищный кредит'), 'Тип_кредита']=0
data2.loc[(data2['Продукт']=='Кредит на предпринимательскую деятельность') |
          (data2['Продукт']=='Энергосберегающие технологии'), 'Тип_кредита']=1

In [24]:
data2 = data2.rename(columns={'Залог (группа-поручительство, категория 1 - недвижимость, категория 2-движимое имущество, категория 3-поручительство, категория 4 - без залога)':'Залог'})

In [25]:
data2.isnull().sum()

Married                                               0
Gender                                                0
Nationality                                           0
Educ                                                  0
FamilySize                                            0
BusExper                                          20352
Filial                                                0
Region                                                0
Навпреление деятельности                              0
Продукт                                              86
Сектор                                                0
Валюта кредита                                        0
Сумма кредита                                         0
Срок кредита                                          0
Ежемесячная сумма погашения по графику                0
Льготный период (месяц)                            1515
Залог                                                 0
Обязательства                                   

In [26]:
data2.describe()

Gender    FamilySize  Сумма кредита  Срок кредита  \
count  85422.000000  85422.000000   8.542200e+04  85422.000000   
mean       0.408326      2.675774   4.762151e+03     12.763843   
std        0.491527      1.832014   1.178836e+04      5.396642   
min        0.000000      0.000000   2.000000e+01      1.000000   
25%        0.000000      1.000000   1.400000e+03     12.000000   
50%        0.000000      3.000000   3.000000e+03     12.000000   
75%        1.000000      4.000000   5.000000e+03     12.000000   
max        1.000000      8.000000   2.000000e+06    120.000000   

       Ежемесячная сумма погашения по графику  Льготный период (месяц)  \
count                            85422.000000             83907.000000   
mean                               582.871092                 1.901033   
std                               1292.947867                 2.698440   
min                                  2.000000                 0.000000   
25%                                164.000000                 0.000000   
50%                                333.000000                 1.000000   
75%                                632.000000                 3.000000   
max                             196947.000000                24.000000   

       Обязательства  Дни просрочки  Количество просрочек  \
count   85383.000000   85422.000000          85422.000000   
mean      702.061148      15.161340              1.869483   
std      4213.904780      48.960655              2.983956   
min         0.000000       0.000000              0.000000   
25%         0.000000       0.000000              0.000000   
50%         0.000000       1.000000              1.000000   
75%        77.000000       7.000000              2.000000   
max    266840.000000     913.000000             26.000000   

       Этап кредитования (который раз получает кредит  Чистая прибыль  \
count                                    85422.000000    79707.000000   
mean                                         2.750111     2076.684115   
std                                          3.202474     5694.100376   
min                                          0.000000   -70150.000000   
25%                                          1.000000      560.000000   
50%                                          2.000000     1050.000000   
75%                                          3.000000     1861.000000   
max                                         81.000000   595504.000000   

                Age  FamilyStatus     Education  business_experience  \
count  85422.000000  85422.000000  85422.000000         65069.000000   
mean      40.076198      0.222765      3.322130             1.271619   
std       11.338563      0.633192      1.173438             1.443552   
min       14.000000      0.000000      0.000000             0.000000   
25%       31.000000      0.000000      2.000000             0.000000   
50%       39.000000      0.000000      4.000000             1.000000   
75%       49.000000      0.000000      4.000000             2.000000   
max       82.000000      3.000000      5.000000             7.000000   

        Тип_кредита  
count  85336.000000  
mean       0.681154  
std        0.466032  
min        0.000000  
25%        0.000000  
50%        1.000000  
75%        1.000000  
max        1.000000

In [27]:
from fancyimpute import IterativeImputer
import pandas as pd

# Список столбцов для предсказания
columns_to_fill = ['Льготный период (месяц)', 'Обязательства', 'Чистая прибыль', 'Тип_кредита','business_experience']

# Проходим по всем столбцам для проверки корреляции
for column in columns_to_fill:
    # Проверяем корреляцию только для столбцов с числовыми данными
    if data2[column].dtype in ['int64', 'float64']:
        # Выбираем строки, где нет пропущенных значений в текущем столбце
        valid_data = data2[columns_to_fill + [column]].dropna()

        # Проверяем, что в данных для обучения модели нет пропущенных значений
        if not valid_data.isnull().values.any():
            # Используем fancyimpute для итеративной импутации (заполнения)
            imputer = IterativeImputer(sample_posterior=True, max_iter=10, random_state=0)
            imputed_values = imputer.fit_transform(data2[columns_to_fill + [column]])

            mask = data2[column].isnull()
            data2.loc[mask, column] = imputed_values[mask, -1]

In [28]:
#Все значения столбцов делаем целочисенным после импутации
data2[['Льготный период (месяц)', 'Тип_кредита','business_experience']] = data2[['Льготный период (месяц)', 'Тип_кредита','business_experience']].astype(int)

In [29]:
data2.isnull().sum()

Married                                               0
Gender                                                0
Nationality                                           0
Educ                                                  0
FamilySize                                            0
BusExper                                          20352
Filial                                                0
Region                                                0
Навпреление деятельности                              0
Продукт                                              86
Сектор                                                0
Валюта кредита                                        0
Сумма кредита                                         0
Срок кредита                                          0
Ежемесячная сумма погашения по графику                0
Льготный период (месяц)                               0
Залог                                                 0
Обязательства                                   

In [30]:
data2.describe()

Gender    FamilySize  Сумма кредита  Срок кредита  \
count  85422.000000  85422.000000   8.542200e+04  85422.000000   
mean       0.408326      2.675774   4.762151e+03     12.763843   
std        0.491527      1.832014   1.178836e+04      5.396642   
min        0.000000      0.000000   2.000000e+01      1.000000   
25%        0.000000      1.000000   1.400000e+03     12.000000   
50%        0.000000      3.000000   3.000000e+03     12.000000   
75%        1.000000      4.000000   5.000000e+03     12.000000   
max        1.000000      8.000000   2.000000e+06    120.000000   

       Ежемесячная сумма погашения по графику  Льготный период (месяц)  \
count                            85422.000000             85422.000000   
mean                               582.871092                 1.885053   
std                               1292.947867                 2.677047   
min                                  2.000000                 0.000000   
25%                                164.000000                 0.000000   
50%                                333.000000                 1.000000   
75%                                632.000000                 3.000000   
max                             196947.000000                24.000000   

       Обязательства  Дни просрочки  Количество просрочек  \
count   85422.000000   85422.000000          85422.000000   
mean      702.061148      15.161340              1.869483   
std      4212.942715      48.960655              2.983956   
min         0.000000       0.000000              0.000000   
25%         0.000000       0.000000              0.000000   
50%         0.000000       1.000000              1.000000   
75%        80.000000       7.000000              2.000000   
max    266840.000000     913.000000             26.000000   

       Этап кредитования (который раз получает кредит  Чистая прибыль  \
count                                    85422.000000    85422.000000   
mean                                         2.750111     2076.684115   
std                                          3.202474     5500.324352   
min                                          0.000000   -70150.000000   
25%                                          1.000000      595.000000   
50%                                          2.000000     1130.000000   
75%                                          3.000000     2076.684098   
max                                         81.000000   595504.000000   

                Age  FamilyStatus     Education  business_experience  \
count  85422.000000  85422.000000  85422.000000         85422.000000   
mean      40.076198      0.222765      3.322130             1.206902   
std       11.338563      0.633192      1.173438             1.265196   
min       14.000000      0.000000      0.000000             0.000000   
25%       31.000000      0.000000      2.000000             0.000000   
50%       39.000000      0.000000      4.000000             1.000000   
75%       49.000000      0.000000      4.000000             2.000000   
max       82.000000      3.000000      5.000000             7.000000   

        Тип_кредита  
count  85422.000000  
mean       0.680469  
std        0.466298  
min        0.000000  
25%        0.000000  
50%        1.000000  
75%        1.000000  
max        1.000000

Судя по описательной статистике до и после импутации можно сказать, что множественная импутация заполнила значения довольно хорошо, так как сильных разбросов в значениях описательной статистике не видно

In [31]:
# Инженерия признаков
data2['Сред_просрочки'] = data2['Дни просрочки'] / data2['Количество просрочек']
data2['has_overdue'] = (data2['Дни просрочки'] > 0).astype(int)
data2['high_debt'] = (data2['Обязательства'] > 10000).astype(int)


In [32]:
#Заполняем пустые ячейки значением 0, так как клиент ни разу не просрочивал
data2['Сред_просрочки'].fillna(0, inplace=True)
data2['Сред_просрочки'] = data2['Сред_просрочки'].astype(int)

In [33]:
data2.head()

Married  Gender Nationality           Educ  FamilySize BusExper  \
0  Оиладор       0       Узбек            Оли           3       15   
1  Оиладор       0       Точик          Миёна           3      NaN   
2  Оиладор       0       Точик          Миёна           3      NaN   
3   Беоила       1       Точик  Миёнаи махсус           0        3   
4  Оиладор       0       Узбек          Миёна           1        5   

       Filial         Region   Навпреление деятельности  \
0  Истаравшан      Шахристон  Чорводори ва коркарди шир   
1  Истаравшан      Гули сурх       Приобретение техники   
2  Истаравшан      Гули сурх                Ремонт дома   
3  Истаравшан      Шахристон  Чорводори ва коркарди шир   
4      Хучанд  Худжанд-Центр  Чорводори ва коркарди шир   

                                      Продукт  ...  \
0  Кредит на предпринимательскую деятельность  ...   
1                      Потребительский кредит  ...   
2                      Потребительский кредит  ...   
3  Кредит на предпринимательскую деятельность  ...   
4  Кредит на предпринимательскую деятельность  ...   

  Этап кредитования (который раз получает кредит Чистая прибыль  Age  \
0                                              1          180.0   39   
1                                              1            0.0   80   
2                                              2         1648.0   82   
3                                              1          150.0   80   
4                                              2          280.0   75   

   FamilyStatus  Education  business_experience Тип_кредита  Сред_просрочки  \
0             0          1                    2           1              11   
1             0          4                    1           0               2   
2             0          4                    1           0               0   
3             1          2                    0           1               0   
4             0          4                    0           1               7   

   has_overdue  high_debt  
0            1          0  
1            1          0  
2            0          0  
3            0          0  
4            1          0  

[5 rows x 30 columns]

In [34]:
data2.columns

Index(['Married', 'Gender', 'Nationality', 'Educ', 'FamilySize', 'BusExper',
       'Filial', 'Region', 'Навпреление деятельности', 'Продукт', 'Сектор',
       'Валюта кредита', 'Сумма кредита', 'Срок кредита',
       'Ежемесячная сумма погашения по графику', 'Льготный период (месяц)',
       'Залог', 'Обязательства', 'Дни просрочки', 'Количество просрочек',
       'Этап кредитования (который раз получает кредит', 'Чистая прибыль',
       'Age', 'FamilyStatus', 'Education', 'business_experience',
       'Тип_кредита', 'Сред_просрочки', 'has_overdue', 'high_debt'],
      dtype='object')

In [35]:
#Используем get_dummies, так как у нас нет явного порядка
data2 = pd.get_dummies(data2, columns=['Nationality'], dummy_na=False)

In [36]:
data2

Married  Gender           Educ  FamilySize BusExper      Filial  \
0      Оиладор       0            Оли           3       15  Истаравшан   
1      Оиладор       0          Миёна           3      NaN  Истаравшан   
2      Оиладор       0          Миёна           3      NaN  Истаравшан   
3       Беоила       1  Миёнаи махсус           0        3  Истаравшан   
4      Оиладор       0          Миёна           1        5      Хучанд   
...        ...     ...            ...         ...      ...         ...   
85417  Оиладор       0          Миёна           0       10  Ч. Расулов   
85418   Беоила       0          Миёна           1      NaN  Ч. Расулов   
85419  Оиладор       1            Оли           3      NaN  Ч. Расулов   
85420  Оиладор       0          Миёна           2        5  Ч. Расулов   
85421  Оиладор       1          Миёна           2        4  Ч. Расулов   

              Region   Навпреление деятельности  \
0          Шахристон  Чорводори ва коркарди шир   
1          Гули сурх       Приобретение техники   
2          Гули сурх                Ремонт дома   
3          Шахристон  Чорводори ва коркарди шир   
4      Худжанд-Центр  Чорводори ва коркарди шир   
...              ...                        ...   
85417          Куруш  Чорводори ва коркарди шир   
85418       Спитамен                  Кишту кор   
85419     Гулякандоз                Ремонт дома   
85420     Дж.Расулов  Чорводори ва коркарди шир   
85421     Дж.Расулов                      Савдо   

                                          Продукт  \
0      Кредит на предпринимательскую деятельность   
1                          Потребительский кредит   
2                          Потребительский кредит   
3      Кредит на предпринимательскую деятельность   
4      Кредит на предпринимательскую деятельность   
...                                           ...   
85417  Кредит на предпринимательскую деятельность   
85418  Кредит на предпринимательскую деятельность   
85419                             Жилищный кредит   
85420  Кредит на предпринимательскую деятельность   
85421  Кредит на предпринимательскую деятельность   

                                Сектор  ... high_debt  Nationality_Другие  \
0                       Животноводство  ...         0               False   
1                      Потреб Экспресс  ...         0               False   
2                          Потребнужды  ...         0               False   
3                       Животноводство  ...         0               False   
4                       Животноводство  ...         0               False   
...                                ...  ...       ...                 ...   
85417                   Животноводство  ...         0               False   
85418           Сельхозкультура (ТАФФ)  ...         0               False   
85419  KFW - Покупка и строит-во жилья  ...         1               False   
85420                   Животноводство  ...         0               False   
85421                         Торговля  ...         0               False   

       Nationality_Карачои  Nationality_Карис  Nationality_Киргиз  \
0                    False              False               False   
1                    False              False               False   
2                    False              False               False   
3                    False              False               False   
4                    False              False               False   
...                    ...                ...                 ...   
85417                False              False               False   
85418                False              False               False   
85419                False              False               False   
85420                False              False               False   
85421                False              False               False   

      Nationality_Рус  Nationality_Тотор  Nationality_Точик  \
0               False              False           

In [37]:
#Используем get_dummies, так как у нас нет явного порядка
data2 = pd.get_dummies(data2, columns=['Filial'], dummy_na=False)

In [38]:
#Используем get_dummies, так как у нас нет явного порядка
data2 = pd.get_dummies(data2, columns=['Region'], dummy_na=False)

In [39]:
data2.head()

Married  Gender           Educ  FamilySize BusExper  \
0  Оиладор       0            Оли           3       15   
1  Оиладор       0          Миёна           3      NaN   
2  Оиладор       0          Миёна           3      NaN   
3   Беоила       1  Миёнаи махсус           0        3   
4  Оиладор       0          Миёна           1        5   

    Навпреление деятельности                                     Продукт  \
0  Чорводори ва коркарди шир  Кредит на предпринимательскую деятельность   
1       Приобретение техники                      Потребительский кредит   
2                Ремонт дома                      Потребительский кредит   
3  Чорводори ва коркарди шир  Кредит на предпринимательскую деятельность   
4  Чорводори ва коркарди шир  Кредит на предпринимательскую деятельность   

            Сектор Валюта кредита  Сумма кредита  ...  Region_Турсунзода  \
0   Животноводство     Доллар США         1000.0  ...              False   
1  Потреб Экспресс         Сомони         1000.0  ...              False   
2      Потребнужды         Сомони         7600.0  ...              False   
3   Животноводство         Сомони         1500.0  ...              False   
4   Животноводство         Сомони         2500.0  ...              False   

   Region_Уяс  Region_Файзобод Region_Х.Алиев  Region_Хисор  \
0       False            False          False         False   
1       False            False          False         False   
2       False            False          False         False   
3       False            False          False         False   
4       False            False          False         False   

   Region_Худжанд-Панчшанбе  Region_Худжанд-Центр  Region_Чашмасор  \
0                     False                 False            False   
1                     False                 False            False   
2                     False                 False            False   
3                     False                 False            False   
4                     False                  True            False   

   Region_Шарк  Region_Шахристон  
0        False              True  
1        False             False  
2        False             False  
3        False              True  
4        False             False  

[5 rows x 95 columns]

In [40]:
#Убираем одинаковые по смыслу значения и переименовываем все на русском
data2.loc[(data2['Навпреление деятельности']=='Чорводори ва коркарди шир') | 
          (data2['Навпреление деятельности']=='Коркарди шир') | (data2['Навпреление деятельности']=='Животноводство'),
         'Навпреление деятельности']='Животноводство и переработка молока'
data2.loc[data2['Навпреление деятельности']=='Савдо', 'Навпреление деятельности']='торговля'
data2.loc[data2['Навпреление деятельности']=='Кишту кор', 'Навпреление деятельности']='Земледелие'
data2.loc[data2['Навпреление деятельности']=='Хидматрасони', 'Навпреление деятельности']='Услуги'
data2.loc[data2['Навпреление деятельности']=='Истехсолот', 'Навпреление деятельности']='Производство'
data2.loc[data2['Навпреление деятельности']=='Кишоварзи', 'Навпреление деятельности']='Сельское хозяйство'
data2.loc[(data2['Навпреление деятельности']=='Мевахушккуни') | 
          (data2['Навпреление деятельности']=='Сушка фрутов'), 'Навпреление деятельности']='Сушка фруктов'

In [41]:
data2 = data2.rename(columns={'Навпреление деятельности':'Цель_кредита'})

In [42]:
#Используем get_dummies, так как у нас нет явного порядка
data2 = pd.get_dummies(data2, columns=['Цель_кредита'], dummy_na=False)

In [43]:
#переименовываем все на русском
data2.loc[data2['Сектор']=='Тахсил', 'Сектор']='Образование'
data2.loc[data2['Сектор']=='Комьёби', 'Сектор']='Достижения'

In [44]:
#Используем get_dummies, так как у нас нет явного порядка
data2 = pd.get_dummies(data2, columns=['Сектор'], dummy_na=False)

In [45]:
data2.head()

Married  Gender           Educ  FamilySize BusExper  \
0  Оиладор       0            Оли           3       15   
1  Оиладор       0          Миёна           3      NaN   
2  Оиладор       0          Миёна           3      NaN   
3   Беоила       1  Миёнаи махсус           0        3   
4  Оиладор       0          Миёна           1        5   

                                      Продукт Валюта кредита  Сумма кредита  \
0  Кредит на предпринимательскую деятельность     Доллар США         1000.0   
1                      Потребительский кредит         Сомони         1000.0   
2                      Потребительский кредит         Сомони         7600.0   
3  Кредит на предпринимательскую деятельность         Сомони         1500.0   
4  Кредит на предпринимательскую деятельность         Сомони         2500.0   

   Срок кредита  Ежемесячная сумма погашения по графику  ...  \
0            12                                  181.00  ...   
1            12                                  109.67  ...   
2            12                                  805.00  ...   
3            12                                  162.93  ...   
4            12                                  465.94  ...   

   Сектор_Производство Сектор_Ремонт жилья  Сектор_Сельхозкультура (Сароб)  \
0                False               False                           False   
1                False               False                           False   
2                False               False                           False   
3                False               False                           False   
4                False               False                           False   

   Сектор_Сельхозкультура (ТАФФ)  Сектор_Сельхозкультура-кредитная линия  \
0                          False                                   False   
1                          False                                   False   
2                          False                                   False   
3                          False                                   False   
4                          False                                   False   

   Сектор_Сельхозтехника (ТАФФ)  Сектор_Старт-бизнес  Сектор_Товары в кредит  \
0                         False                False                   False   
1                         False                False                   False   
2                         False                False                   False   
3                         False                False                   False   
4                         False                False                   False   

   Сектор_Торговля  Сектор_Услуги  
0            False          False  
1            False          False  
2            False          False  
3            False          False  
4            False          False  

[5 rows x 140 columns]

In [46]:
data2 = pd.get_dummies(data2, columns=['Валюта кредита'])

In [47]:
data2 = pd.get_dummies(data2, columns=['Залог'])

In [48]:
data2.head()

Married  Gender           Educ  FamilySize BusExper  \
0  Оиладор       0            Оли           3       15   
1  Оиладор       0          Миёна           3      NaN   
2  Оиладор       0          Миёна           3      NaN   
3   Беоила       1  Миёнаи махсус           0        3   
4  Оиладор       0          Миёна           1        5   

                                      Продукт  Сумма кредита  Срок кредита  \
0  Кредит на предпринимательскую деятельность         1000.0            12   
1                      Потребительский кредит         1000.0            12   
2                      Потребительский кредит         7600.0            12   
3  Кредит на предпринимательскую деятельность         1500.0            12   
4  Кредит на предпринимательскую деятельность         2500.0            12   

   Ежемесячная сумма погашения по графику  Льготный период (месяц)  ...  \
0                                  181.00                        6  ...   
1                                  109.67                        0  ...   
2                                  805.00                        0  ...   
3                                  162.93                        1  ...   
4                                  465.94                        6  ...   

   Сектор_Торговля  Сектор_Услуги  Валюта кредита_Доллар США  \
0            False          False                       True   
1            False          False                      False   
2            False          False                      False   
3            False          False                      False   
4            False          False                      False   

   Валюта кредита_Рос.рубль  Валюта кредита_Сомони  Залог_Группа  \
0                     False                  False         False   
1                     False                   True         False   
2                     False                   True         False   
3                     False                   True         False   
4                     False                   True         False   

   Залог_Категория 1  Залог_Категория 2  Залог_Категория 3  Залог_Категория 4  
0              False               True              False              False  
1              False              False              False               True  
2              False              False               True              False  
3              False               True              False              False  
4              False              False              False               True  

[5 rows x 146 columns]

In [49]:
data2 = data2.rename(columns={'Этап кредитования (который раз получает кредит':'Количество_кредитов'})

In [50]:
data2=data2.drop(['Married','Educ','BusExper','Продукт'], axis=1)

Пункт разведочный анализ приведен в файле Arvand.ipynb

# Используйте всевозможные входные параметры (независимые переменные) и зависимую переменную «Дни просрочки»

Я буду решать задачу классификации с выбором трешхолда = 20 дней

In [51]:
data2.head()

Gender  FamilySize  Сумма кредита  Срок кредита  \
0       0           3         1000.0            12   
1       0           3         1000.0            12   
2       0           3         7600.0            12   
3       1           0         1500.0            12   
4       0           1         2500.0            12   

   Ежемесячная сумма погашения по графику  Льготный период (месяц)  \
0                                  181.00                        6   
1                                  109.67                        0   
2                                  805.00                        0   
3                                  162.93                        1   
4                                  465.94                        6   

   Обязательства  Дни просрочки  Количество просрочек  Количество_кредитов  \
0            0.0             46                     4                    1   
1            0.0             11                     5                    1   
2            0.0              0                     0                    2   
3            0.0              0                     0                    1   
4          543.0             15                     2                    2   

   ...  Сектор_Торговля  Сектор_Услуги  Валюта кредита_Доллар США  \
0  ...            False          False                       True   
1  ...            False          False                      False   
2  ...            False          False                      False   
3  ...            False          False                      False   
4  ...            False          False                      False   

   Валюта кредита_Рос.рубль  Валюта кредита_Сомони  Залог_Группа  \
0                     False                  False         False   
1                     False                   True         False   
2                     False                   True         False   
3                     False                   True         False   
4                     False                   True         False   

   Залог_Категория 1  Залог_Категория 2  Залог_Категория 3  Залог_Категория 4  
0              False               True              False              False  
1              False              False              False               True  
2              False              False               True              False  
3              False               True              False              False  
4              False              False              False               True  

[5 rows x 142 columns]

In [52]:
data2.loc[data2['Сред_просрочки']<=20, 'выдача']=1
data2.loc[data2['Сред_просрочки']>20, 'выдача']=0
data2 = data2.drop('Сред_просрочки', axis=1)
data2

Gender  FamilySize  Сумма кредита  Срок кредита  \
0           0           3         1000.0            12   
1           0           3         1000.0            12   
2           0           3         7600.0            12   
3           1           0         1500.0            12   
4           0           1         2500.0            12   
...       ...         ...            ...           ...   
85417       0           0         5000.0            12   
85418       0           1         1400.0            24   
85419       1           3        30000.0            36   
85420       0           2         3000.0            12   
85421       1           2         8000.0            12   

       Ежемесячная сумма погашения по графику  Льготный период (месяц)  \
0                                      181.00                        6   
1                                      109.67                        0   
2                                      805.00                        0   
3                                      162.93                        1   
4                                      465.94                        6   
...                                       ...                      ...   
85417                                  558.00                        1   
85418                                   74.00                        0   
85419                                 1264.00                        1   
85420                                  295.00                        0   
85421                                  888.00                        1   

       Обязательства  Дни просрочки  Количество просрочек  \
0                0.0             46                     4   
1                0.0             11                     5   
2                0.0              0                     0   
3                0.0              0                     0   
4              543.0             15                     2   
...              ...            ...                   ...   
85417            0.0             53                     6   
85418            0.0              0                     0   
85419        44800.0              0                     0   
85420            0.0              2                     2   
85421         1200.0              0                     0   

       Количество_кредитов  ...  Сектор_Услуги  Валюта кредита_Доллар США  \
0                        1  ...          False                       True   
1                        1  ...          False                      False   
2                        2  ...          False                      False   
3                        1  ...          False                      False   
4                        2  ...          False                      False   
...                    ...  ...            ...                        ...   
85417                    1  ...          False                      False   
85418                    1  ...          False                       True   
85419                    1  ...          False                      False   
85420                    1  ...          False                      False   
85421                    6  ...          False                      False   

       Валюта кредита_Рос.рубль  Валюта кредита_Сомони  Залог_Группа  \
0                         False                  False         False   
1                         False                   True         False   
2                         False                   True         False   
3                         False                   True         False   
4                         False                   True         False   
...                         ...                    ...           ...   
85417                     False                   True         False   
85418                     False                  False         False   
85419                     False                   True         False   
85420                     False            

In [52]:
data2['выдача'].value_counts()

выдача
1.0    83704
0.0     1718
Name: count, dtype: int64

In [53]:
from sklearn.model_selection import train_test_split
X = data2.drop(["выдача", "Дни просрочки","Количество просрочек"], axis=1)
y = data2["выдача"]
# Разделение данных на тренировочный и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [54]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from xgboost import XGBClassifier
from sklearn.ensemble import GradientBoostingClassifier, RandomForestClassifier
from catboost import CatBoostClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split, GridSearchCV

### А) логическая регрессия

In [55]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.metrics import roc_auc_score

In [56]:
from sklearn.linear_model import LogisticRegression
# Инициализация модели логистической регрессии
logistic_regression = LogisticRegression(random_state=0, class_weight='balanced')
# Определение сетки параметров для поиска
param_grid_logistic = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],
    'penalty': ['l1', 'l2']
}

# Поиск по сетке с использованием кросс-валидации
grid_search_logistic = GridSearchCV(logistic_regression, param_grid_logistic, cv=5, scoring='accuracy')
grid_search_logistic.fit(X_train, y_train)

# Получение лучших параметров
best_logistic = grid_search_logistic.best_estimator_


# Обучение модели на тренировочных данных
best_logistic.fit(X_train, y_train)

# Предсказание на тестовых данных
y_pred = best_logistic.predict(X_test)

# Оценка качества модели
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)

# Вывод результатов

print(f'Accuracy: {accuracy}')
print(f'Confusion Matrix:\n{conf_matrix}')
print(f'Classification Report:\n{classification_rep}')

C:\Users\extre\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\extre\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_

Accuracy: 0.6650017559605104
Confusion Matrix:
[[  386   136]
 [ 8449 16656]]
Classification Report:
              precision    recall  f1-score   support

         0.0       0.04      0.74      0.08       522
         1.0       0.99      0.66      0.80     25105

    accuracy                           0.67     25627
   macro avg       0.52      0.70      0.44     25627
weighted avg       0.97      0.67      0.78     25627



C:\Users\extre\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [57]:
auc_lr = roc_auc_score(y_test, best_logistic.predict_proba(X_test)[:, 1])
best_params_logistic = grid_search_logistic.best_params_
print("Best Parameters for Logistic Regression:", best_params_logistic)
print("Логическая регрессия:")
print(f'AUC: {auc_lr}')

Best Parameters for Logistic Regression: {'C': 1, 'penalty': 'l2'}
Логическая регрессия:
AUC: 0.764641227152473


### Б) дерево решений;

In [58]:
from sklearn.tree import DecisionTreeClassifier
# Инициализация модели дерева решений
decision_tree = DecisionTreeClassifier(random_state=42, class_weight='balanced')

param_grid_dt = {
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}
grid_search_dt = GridSearchCV(decision_tree, param_grid_dt, cv=5, scoring='accuracy')
grid_search_dt.fit(X_train, y_train)
best_dt = grid_search_dt.best_estimator_
# Обучение модели на тренировочных данных
best_dt.fit(X_train, y_train)

# Предсказание на тестовых данных
y_pred = best_dt.predict(X_test)

# Оценка качества модели
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)

# Вывод результатов

print(f'Accuracy: {accuracy}')
print(f'Confusion Matrix:\n{conf_matrix}')
print(f'Classification Report:\n{classification_rep}')

Accuracy: 0.9623834237327819
Confusion Matrix:
[[   58   464]
 [  500 24605]]
Classification Report:
              precision    recall  f1-score   support

         0.0       0.10      0.11      0.11       522
         1.0       0.98      0.98      0.98     25105

    accuracy                           0.96     25627
   macro avg       0.54      0.55      0.54     25627
weighted avg       0.96      0.96      0.96     25627



In [59]:
best_params_dt = grid_search_dt.best_params_
print("Best Parameters for Decision Tree:", best_params_dt)
auc_DecTree = roc_auc_score(y_test, best_dt.predict_proba(X_test)[:, 1])
print("\nдерево решений:")
print(f'AUC: {auc_DecTree}')

Best Parameters for Decision Tree: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2}

дерево решений:
AUC: 0.5455973798933369


### В) случайны лес;

In [60]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

# 7. Подбор гиперпараметров
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

rf = RandomForestClassifier(random_state=42, class_weight='balanced')
grid_search = GridSearchCV(rf, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, y_train)

best_rf = grid_search.best_estimator_

# 8. Обучение модели
best_rf.fit(X_train, y_train)

# 9. Оценка модели
y_pred = best_rf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.4f}")

Accuracy: 0.9794


In [61]:
best_params_rf = grid_search.best_params_
print("Best Parameters for RandomForest:", best_params_rf)

Best Parameters for RandomForest: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 50}


In [62]:
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)
print(f'Accuracy: {accuracy}')
print(f'Confusion Matrix:\n{conf_matrix}')
print(f'Classification Report:\n{classification_rep}')

Accuracy: 0.9793577086666406
Confusion Matrix:
[[    5   517]
 [   12 25093]]
Classification Report:
              precision    recall  f1-score   support

         0.0       0.29      0.01      0.02       522
         1.0       0.98      1.00      0.99     25105

    accuracy                           0.98     25627
   macro avg       0.64      0.50      0.50     25627
weighted avg       0.97      0.98      0.97     25627



In [63]:
auc_randf = roc_auc_score(y_test, best_rf.predict_proba(X_test)[:, 1])
print("\nслучайны лес:")
print(f'AUC: {auc_randf}')


случайны лес:
AUC: 0.828174731262796


### Г) К-ближайших соседей;

In [64]:
from sklearn.neighbors import KNeighborsClassifier

# Инициализация модели k-ближайших соседей
k_neighbors = KNeighborsClassifier()
param_grid_knn = {
    'n_neighbors': [3, 5, 7, 9],
    'weights': ['uniform', 'distance']
}
grid_search_knn = GridSearchCV(k_neighbors, param_grid_knn, cv=5, scoring='accuracy')
grid_search_knn.fit(X_train, y_train)
best_knn = grid_search_knn.best_estimator_
# Обучение модели на тренировочных данных
best_knn.fit(X_train, y_train)

# Предсказание на тестовых данных
y_pred = best_knn.predict(X_test)

# Оценка качества модели
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)

# Вывод результатов
print(f'Accuracy: {accuracy}')
print(f'Confusion Matrix:\n{conf_matrix}')
print(f'Classification Report:\n{classification_rep}')

Accuracy: 0.9796308580793694
Confusion Matrix:
[[    0   522]
 [    0 25105]]
Classification Report:
              precision    recall  f1-score   support

         0.0       0.00      0.00      0.00       522
         1.0       0.98      1.00      0.99     25105

    accuracy                           0.98     25627
   macro avg       0.49      0.50      0.49     25627
weighted avg       0.96      0.98      0.97     25627



C:\Users\extre\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\extre\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\extre\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [65]:
best_params_knn = grid_search_knn.best_params_
print("Best Parameters for KNN:", best_params_knn)
auc_knn = roc_auc_score(y_test, best_knn.predict_proba(X_test)[:, 1])
print("\nK-ближайших соседей:")
print(f'AUC: {auc_knn}')

Best Parameters for KNN: {'n_neighbors': 9, 'weights': 'uniform'}

K-ближайших соседей:
AUC: 0.5550238805446245


### Е) алгоритмы бустинга

In [66]:
from xgboost import XGBClassifier
xgb = XGBClassifier(random_state=42, scale_pos_weight=(y_train == 0).sum() / (y_train == 1).sum())
param_grid_xgb = {
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2]
}
grid_search_xgb = GridSearchCV(xgb, param_grid_xgb, cv=5, scoring='accuracy')
grid_search_xgb.fit(X_train, y_train)
best_xgb = grid_search_xgb.best_estimator_

# Обучение модели на тренировочных данных
best_xgb.fit(X_train, y_train)

# Предсказание на тестовых данных
y_pred_xgb = best_xgb.predict(X_test)

# Оценка качества модели
accuracy_xgb = accuracy_score(y_test, y_pred_xgb)
conf_matrix_xgb = confusion_matrix(y_test, y_pred_xgb)
classification_rep_xgb = classification_report(y_test, y_pred_xgb)

# Вывод результатов
print("XGBoost:")
print(f'Accuracy: {accuracy_xgb}')
print(f'Confusion Matrix:\n{conf_matrix_xgb}')
print(f'Classification Report:\n{classification_rep_xgb}')

C:\Users\extre\anaconda3\Lib\site-packages\xgboost\data.py:335: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(dtype):
C:\Users\extre\anaconda3\Lib\site-packages\xgboost\data.py:338: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead
  is_categorical_dtype(dtype) or is_pa_ext_categorical_dtype(dtype)
C:\Users\extre\anaconda3\Lib\site-packages\xgboost\data.py:384: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead
  if is_categorical_dtype(dtype):
C:\Users\extre\anaconda3\Lib\site-packages\xgboost\data.py:359: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead
  return is_int or is_bool or is_float or is_categorical_dtype(dtype)
C

XGBoost:
Accuracy: 0.8840285636243025
Confusion Matrix:
[[  303   219]
 [ 2753 22352]]
Classification Report:
              precision    recall  f1-score   support

         0.0       0.10      0.58      0.17       522
         1.0       0.99      0.89      0.94     25105

    accuracy                           0.88     25627
   macro avg       0.54      0.74      0.55     25627
weighted avg       0.97      0.88      0.92     25627



In [67]:
best_params_xgb = grid_search_xgb.best_params_
print("Best Parameters for XGBoost:", best_params_xgb)
auc_xgb = roc_auc_score(y_test, best_xgb.predict_proba(X_test)[:, 1])
print("XGBoost:")
print(f'AUC: {auc_xgb}')

Best Parameters for XGBoost: {'learning_rate': 0.2, 'max_depth': 7, 'n_estimators': 200}


C:\Users\extre\anaconda3\Lib\site-packages\xgboost\data.py:335: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(dtype):
C:\Users\extre\anaconda3\Lib\site-packages\xgboost\data.py:338: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead
  is_categorical_dtype(dtype) or is_pa_ext_categorical_dtype(dtype)
C:\Users\extre\anaconda3\Lib\site-packages\xgboost\data.py:384: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead
  if is_categorical_dtype(dtype):
C:\Users\extre\anaconda3\Lib\site-packages\xgboost\data.py:359: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead
  return is_int or is_bool or is_float or is_categorical_dtype(dtype)


XGBoost:
AUC: 0.8802408428660926


In [68]:
from sklearn.ensemble import GradientBoostingClassifier
# Gradient Boosting
gradient_boost = GradientBoostingClassifier(random_state=42)
param_grid_gb = {
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2]
}
grid_search_gb = GridSearchCV(gradient_boost, param_grid_gb, cv=5, scoring='accuracy')
grid_search_gb.fit(X_train, y_train)
best_gb = grid_search_gb.best_estimator_
# Обучение модели на тренировочных данных
best_gb.fit(X_train, y_train)

# Предсказание на тестовых данных
y_pred_gb = best_gb.predict(X_test)

# Оценка качества модели
accuracy_gb = accuracy_score(y_test, y_pred_gb)
conf_matrix_gb = confusion_matrix(y_test, y_pred_gb)
classification_rep_gb = classification_report(y_test, y_pred_gb)

# Вывод результатов
print("\nGradient Boosting:")
print(f'Accuracy: {accuracy_gb}')
print(f'Confusion Matrix:\n{conf_matrix_gb}')
print(f'Classification Report:\n{classification_rep_gb}')



Gradient Boosting:
Accuracy: 0.9796308580793694
Confusion Matrix:
[[    0   522]
 [    0 25105]]
Classification Report:
              precision    recall  f1-score   support

         0.0       0.00      0.00      0.00       522
         1.0       0.98      1.00      0.99     25105

    accuracy                           0.98     25627
   macro avg       0.49      0.50      0.49     25627
weighted avg       0.96      0.98      0.97     25627



C:\Users\extre\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\extre\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\extre\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [69]:
best_params_gb = grid_search_gb.best_params_
print("Best Parameters for GradientBoosting:", best_params_gb)
auc_gb = roc_auc_score(y_test, best_gb.predict_proba(X_test)[:, 1])
print("\nGradient Boosting:")
print(f'AUC: {auc_gb}')

Best Parameters for GradientBoosting: {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 50}

Gradient Boosting:
AUC: 0.8503682617298534


In [70]:
from catboost import CatBoostClassifier
# CatBoost
catboost = CatBoostClassifier(random_state=42, class_weights=[1, (y_train == 0).sum() / (y_train == 1).sum()])
param_grid_cat = {
    'iterations': [50, 100, 200],
    'depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2]
}
grid_search_cat = GridSearchCV(catboost, param_grid_cat, cv=5, scoring='accuracy')
grid_search_cat.fit(X_train, y_train)
best_cat = grid_search_cat.best_estimator_


# Обучение моделей на тренировочных данных
best_cat.fit(X_train, y_train)

# Предсказание на тестовых данных
y_pred_cat = best_cat.predict(X_test)

# Оценка качества модели
accuracy_cat = accuracy_score(y_test, y_pred_cat)
conf_matrix_cat = confusion_matrix(y_test, y_pred_cat)
classification_rep_cat = classification_report(y_test, y_pred_cat)

# Вывод результатов
print("\nCatBoost:")
print(f'Accuracy: {accuracy_cat}')
print(f'Confusion Matrix:\n{conf_matrix_cat}')
print(f'Classification Report:\n{classification_rep_cat}')

0:	learn: 0.6809420	total: 171ms	remaining: 8.39s
1:	learn: 0.6695664	total: 178ms	remaining: 4.28s
2:	learn: 0.6587643	total: 186ms	remaining: 2.92s
3:	learn: 0.6482515	total: 194ms	remaining: 2.23s
4:	learn: 0.6384966	total: 201ms	remaining: 1.81s
5:	learn: 0.6291803	total: 209ms	remaining: 1.53s
6:	learn: 0.6203660	total: 217ms	remaining: 1.33s
7:	learn: 0.6119676	total: 224ms	remaining: 1.18s
8:	learn: 0.6039588	total: 232ms	remaining: 1.06s
9:	learn: 0.5963793	total: 241ms	remaining: 963ms
10:	learn: 0.5891947	total: 250ms	remaining: 886ms
11:	learn: 0.5823899	total: 257ms	remaining: 814ms
12:	learn: 0.5758818	total: 266ms	remaining: 756ms
13:	learn: 0.5697734	total: 273ms	remaining: 702ms
14:	learn: 0.5637943	total: 281ms	remaining: 656ms
15:	learn: 0.5582621	total: 289ms	remaining: 614ms
16:	learn: 0.5530395	total: 298ms	remaining: 578ms
17:	learn: 0.5482268	total: 306ms	remaining: 544ms
18:	learn: 0.5434476	total: 315ms	remaining: 515ms
19:	learn: 0.5390953	total: 324ms	remaini

In [71]:
best_params_cat = grid_search_cat.best_params_
print("Best Parameters for CatBoost:", best_params_cat)
# Вычисление AUC для модели
auc_cat = roc_auc_score(y_test, best_cat.predict_proba(X_test)[:, 1])
# Вывод результатов
print("\nCatBoost:")
print(f'AUC: {auc_cat}')


Best Parameters for CatBoost: {'depth': 7, 'iterations': 200, 'learning_rate': 0.2}

CatBoost:
AUC: 0.8683444094191369


In [72]:
import pickle

# Сохранение модели в файл
with open('RF_model.pkl', 'wb') as file:
    pickle.dump(best_rf, file)
with open('XGB_model.pkl', 'wb') as file:
    pickle.dump(best_xgb, file)
with open('CatBoost_model.pkl', 'wb') as file:
    pickle.dump(best_cat, file)

Random Forest Classifier, XGBClassifier, и CatBoost показывают хорошие результаты с высокими значениями точности и полноты.

# 4. Теперь, используя кол-во просроченных дней делайте регрессию.

## Линейная регрессия, МНК 

Я использую методы линейной регресси Ridge и Lasso и в зависимости от MSE и R^2 выберу лучшую модель для регрессии. Так же я использую GridSearchCV для поиска лучшего параметра альфа и результатов модели

In [102]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Ridge

y = data2['Дни просрочки']
X = data2.drop(['Дни просрочки',"выдача", 'Количество просрочек'], axis=1)

# Разделение данных на тренировочный и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)
# Задаем параметры для перебора
param_grid = {'alpha': [0.1, 1.0, 10.0]}

# Инициализируем модель
ridge = Ridge()

# Инициализируем GridSearchCV
grid_search = GridSearchCV(ridge, param_grid, cv=5, scoring='neg_mean_squared_error')

# Обучаем GridSearchCV на данных
grid_search.fit(X_train, y_train)

# Получаем лучшие параметры
best_params = grid_search.best_params_

# Используем лучшие параметры для создания финальной модели
best_ridge = Ridge(**best_params)
best_ridge.fit(X_train, y_train)

# Предсказание на тестовых данных
y_pred = best_ridge.predict(X_test)

# Оценка качества модели
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Вывод результатов
print(f'Mean Squared Error: {mse}')
print(f'R^2 Score: {r2}')
print('Лучшие параметры:', best_params)

Mean Squared Error: 2041.1700882001246
R^2 Score: 0.14348395212462028
Лучшие параметры: {'alpha': 1.0}


Как видно из результатов, Ridge очень плохо справилась с задачей и можно увидеть, что самый оптимальным параметром является значение альфа равное 1.0

In [103]:
from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV


lasso = Lasso()
# Параметры для GridSearchCV
param_grid = {'alpha': [0.1, 1, 10, 100]}


grid_search = GridSearchCV(lasso, param_grid, cv=5, scoring='neg_mean_squared_error')


grid_search.fit(X_train, y_train)


best_params = grid_search.best_params_


best_lasso = Lasso(**best_params)
best_lasso.fit(X_train, y_train)

y_pred = best_lasso.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'Mean Squared Error: {mse}')
print(f'R^2 Score: {r2}')
print('Лучшие параметры:', best_params)

Mean Squared Error: 2057.8179201198477
R^2 Score: 0.13649818681088866
Лучшие параметры: {'alpha': 0.1}


Как видно из результатов, метод Lasso справился с задачей чуть хуже, чем Ridge

## Случайный лес

Для моделей случайного леса и градиентного бустинга я сначала установил несколько параметров для GridSearch. Но тренировка модели была очень долгой(оставил на 4 часа, но даже столько времени не хватило). Поэтому я поставил средние значения, но код не поменял

In [104]:
from sklearn.ensemble import RandomForestRegressor

y = data2['Дни просрочки']
X = data2.drop(['Дни просрочки',"выдача", 'Количество просрочек'], axis=1)

# Разделение данных на тренировочный и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

# Задаем сетку параметров для поиска
param_grid_rf = {
    'n_estimators': [20],
    'max_depth': [10],
    'min_samples_split': [5],
    'min_samples_leaf': [4]
}

# Инициализация модели случайного леса
rf = RandomForestRegressor(random_state=42)

# Создаем объект GridSearchCV
grid_search_rf = GridSearchCV(rf, param_grid_rf, cv=5, scoring='neg_mean_squared_error')

# Подгоняем GridSearchCV на тренировочных данных
grid_search_rf.fit(X_train, y_train)

# Получаем лучшую модель
best_rf = grid_search_rf.best_estimator_

# Печатаем лучшие параметры
print("Лучшие параметры для Random Forest:", grid_search_rf.best_params_)

# Предсказание на тестовых данных с лучшей моделью
y_pred_best_rf = best_rf.predict(X_test)

# Оценка качества модели с лучшими параметрами
mse_best_rf = mean_squared_error(y_test, y_pred_best_rf)
r2_best_rf = r2_score(y_test, y_pred_best_rf)

# Вывод результатов
print(f'Mean Squared Error (Random Forest, best params): {mse_best_rf}')
print(f'R^2 Score (Random Forest, best params): {r2_best_rf}')



Лучшие параметры для Random Forest: {'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 5, 'n_estimators': 20}
Mean Squared Error (Random Forest, best params): 1957.824656177172
R^2 Score (Random Forest, best params): 0.17845737274127904


   Mean Squared Error (MSE): 1957.82 - уменьшился по сравнению с результатами линейной регрессии, что указывает на лучшую точность предсказаний модели, но все еще результатыт недостаточно хорошие.

   R^2 Score (Random Forest): 0.178 - увеличился по сравнению с линейной регрессией, что также говорит о более хорошем объяснении изменчивости в данных.

## GradientBoost

In [105]:
from sklearn.ensemble import GradientBoostingRegressor

y = data2['Дни просрочки']
X = data2.drop(['Дни просрочки',"выдача", 'Количество просрочек'], axis=1)

# Разделение данных на тренировочный и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

# Задаем сетку параметров для поиска
param_grid_gb = {
    'n_estimators': [20],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [10],
    'min_samples_split': [5],
    'min_samples_leaf': [4],
    'subsample': [0.8, 1.0]
}

# Инициализация модели GradientBoostingRegressor
gb = GradientBoostingRegressor(random_state=42)

# Создаем объект GridSearchCV
grid_search_gb = GridSearchCV(gb, param_grid_gb, cv=5, scoring='neg_mean_squared_error')

# Подгоняем GridSearchCV на тренировочных данных
grid_search_gb.fit(X_train, y_train)

# Получаем лучшую модель
best_gb = grid_search_gb.best_estimator_

# Печатаем лучшие параметры
print("Лучшие параметры для GradientBoostingRegressor:", grid_search_gb.best_params_)

# Предсказание на тестовых данных с лучшей моделью
y_pred_best_gb = best_gb.predict(X_test)

# Оценка качества модели с лучшими параметрами
mse_best_gb = mean_squared_error(y_test, y_pred_best_gb)
r2_best_gb = r2_score(y_test, y_pred_best_gb)

# Вывод результатов
print(f'Mean Squared Error (GradientBoostingRegressor, best params): {mse_best_gb}')
print(f'R^2 Score (GradientBoostingRegressor, best params): {r2_best_gb}')


Лучшие параметры для GradientBoostingRegressor: {'learning_rate': 0.1, 'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 5, 'n_estimators': 20, 'subsample': 0.8}
Mean Squared Error (GradientBoostingRegressor, best params): 1954.2706557111303
R^2 Score (GradientBoostingRegressor, best params): 0.17994870286164422


   Mean Squared Error (MSE): 1954.27 - это чуть меньше, чем у Random Forest, но все равно модели очень плохо справляются с задачей


In [106]:
import pickle
#Сохраняем модель случайный лес в пикл файл
with open('days.pkl', 'wb') as file:
    pickle.dump(best_gb, file)

### 5) Если клиенту отказывается в кредите, то в какую сумму кредить выдавать?
### 6) Модель не только должна сказать выдать или не выдать, но и вероятность возврата в процентах

Для решения данной задачи нам нужно сделать регрессию суммы кредита только в случаях, когда значение в столбце "выдача" равно 0, то есть когда кредит не выдаётся.

### RandomForestRegressor

In [53]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestRegressor
X_regression = data2[data2['выдача'] == 0].drop(['Дни просрочки','Сумма кредита', 'Количество просрочек'], axis=1)
y_regression = data2[data2['выдача'] == 0]['Сумма кредита']

# Разделение данных на тренировочный и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X_regression,y_regression, test_size=0.3, random_state=0)

# Задаем сетку параметров для поиска
param_grid_rf = {
    'n_estimators': [20, 30],
    'max_depth': [10, 20, 30],
    'min_samples_split': [3, 5, 7],
    'min_samples_leaf': [1, 2, 4]
}

# Инициализация модели случайного леса
rf = RandomForestRegressor(random_state=42)

# Создаем объект GridSearchCV
grid_search_rf = GridSearchCV(rf, param_grid_rf, cv=5, scoring='neg_mean_squared_error')

# Подгоняем GridSearchCV на тренировочных данных
grid_search_rf.fit(X_train, y_train)

# Получаем лучшую модель
best_rf2 = grid_search_rf.best_estimator_

# Печатаем лучшие параметры
print("Лучшие параметры для Random Forest:", grid_search_rf.best_params_)

# Предсказание на тестовых данных с лучшей моделью
y_pred_best_rf = best_rf2.predict(X_test)

# Оценка качества модели с лучшими параметрами
mse_best_rf = mean_squared_error(y_test, y_pred_best_rf)
r2_best_rf = r2_score(y_test, y_pred_best_rf)

# Вывод результатов
print(f'Mean Squared Error (Random Forest, best params): {mse_best_rf}')
print(f'R^2 Score (Random Forest, best params): {r2_best_rf}')



Лучшие параметры для Random Forest: {'max_depth': 20, 'min_samples_leaf': 1, 'min_samples_split': 3, 'n_estimators': 20}
Mean Squared Error (Random Forest, best params): 34159767.14218853
R^2 Score (Random Forest, best params): 0.8259659263391635


### метод Ridge линейной регрессии

In [54]:
from sklearn.linear_model import Ridge
# Задаем параметры для перебора
param_grid = {'alpha': [0.1, 1.0, 10.0, 20.0, 30.0, 40.0, 50.0]}

# Инициализируем модель
ridge2 = Ridge()

# Инициализируем GridSearchCV
grid_search = GridSearchCV(ridge2, param_grid, cv=5, scoring='neg_mean_squared_error')

# Обучаем GridSearchCV на данных
grid_search.fit(X_train, y_train)

# Получаем лучшие параметры
best_params2 = grid_search.best_params_

# Используем лучшие параметры для создания финальной модели
best_ridge2 = Ridge(**best_params2)
best_ridge2.fit(X_train, y_train)

# Предсказание на тестовых данных
y_pred = best_ridge2.predict(X_test)

# Оценка качества модели
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Вывод результатов
print(f'Mean Squared Error: {mse}')
print(f'R^2 Score: {r2}')
print('Лучшие параметры:', best_params2)

Mean Squared Error: 38739667.090947226
R^2 Score: 0.8026326687755563
Лучшие параметры: {'alpha': 50.0}


### метод Lasso линейной регрессии

In [55]:
from sklearn.linear_model import Lasso
lasso2 = Lasso()
# Параметры для GridSearchCV
param_grid = {'alpha': [0.1, 0.2, 1, 10, 100]}


grid_search = GridSearchCV(lasso2, param_grid, cv=5, scoring='neg_mean_squared_error')


grid_search.fit(X_train, y_train)


best_params2 = grid_search.best_params_


best_lasso2 = Lasso(**best_params2)
best_lasso2.fit(X_train, y_train)

y_pred = best_lasso2.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'Mean Squared Error: {mse}')
print(f'R^2 Score: {r2}')
print('Лучшие параметры:', best_params2)

C:\Users\extre\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 3.078e+09, tolerance: 5.947e+06
  model = cd_fast.enet_coordinate_descent(
C:\Users\extre\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.265e+08, tolerance: 1.395e+07
  model = cd_fast.enet_coordinate_descent(
C:\Users\extre\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 8.223e+09, toleranc

Mean Squared Error: 38018074.55829885
R^2 Score: 0.8063089727578788
Лучшие параметры: {'alpha': 100}


### GradientBoostingRegressor

In [56]:
from sklearn.ensemble import GradientBoostingRegressor
# Задаем сетку параметров для поиска
param_grid_gb = {
    'n_estimators': [20, 30, 40],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [10, 20, 30],
    'min_samples_split': [3, 5, 7],
    'min_samples_leaf': [2, 4, 6],
    'subsample': [0.8, 1.0]
}

# Инициализация модели GradientBoostingRegressor
gb = GradientBoostingRegressor(random_state=42)

# Создаем объект GridSearchCV
grid_search_gb = GridSearchCV(gb, param_grid_gb, cv=5, scoring='neg_mean_squared_error')

# Подгоняем GridSearchCV на тренировочных данных
grid_search_gb.fit(X_train, y_train)

# Получаем лучшую модель
best_gb2 = grid_search_gb.best_estimator_

# Печатаем лучшие параметры
print("Лучшие параметры для GradientBoostingRegressor:", grid_search_gb.best_params_)

# Предсказание на тестовых данных с лучшей моделью
y_pred_best_gb = best_gb2.predict(X_test)

# Оценка качества модели с лучшими параметрами
mse_best_gb = mean_squared_error(y_test, y_pred_best_gb)
r2_best_gb = r2_score(y_test, y_pred_best_gb)

# Вывод результатов
print(f'Mean Squared Error (GradientBoostingRegressor, best params): {mse_best_gb}')
print(f'R^2 Score (GradientBoostingRegressor, best params): {r2_best_gb}')


Лучшие параметры для GradientBoostingRegressor: {'learning_rate': 0.1, 'max_depth': 30, 'min_samples_leaf': 6, 'min_samples_split': 3, 'n_estimators': 40, 'subsample': 0.8}
Mean Squared Error (GradientBoostingRegressor, best params): 35740385.44427454
R^2 Score (GradientBoostingRegressor, best params): 0.8179131360238816


По результатам MSE и R^2 видно, что лучше всего с задачей справился модель случайного леса. Но MSE у всех моделей получился очень большим

In [57]:
import pickle
#Сохраняем модель случайный лес в пикл файл
with open('sum.pkl', 'wb') as file:
    pickle.dump(best_rf2, file)